Improve delete_deployment handling

brejoc opened the pull request:
### What does this PR do?

The Kubernetes module function `delete_deployment()` uses `api_instance.delete_namespaced_deployment()` from the Kubernetes lib. This method from the Kubernetes lib returns immediately without giving a success or failure indication, which lets Salt mark the job as failed even though we don't know if it failed or not.
To actually get a result I've implemented a polling via `show_deployment()` to check if the deployment got removed.

If a time limit is hit, we are returning with an error, otherwise it is a success.

I'm not sure about the Windows workaround. Since Windows has no `signal.alarm` implementation, I've fallen back to loop counting. Feedback is very welcome here.


### Previous Behavior

`delete_deployment()` shows failure in any case.


### New Behavior

`delete_deployment()` shows success when deployment got deleted within time limit. Otherwise it shows failure.

### Tests written?

No

brejoc committed:
First simple draft for the deletion verification

brejoc committed:
Added timeout for checking the deployment

If the time limit is hit, the checking is aborted and we return with
return-code None.

brejoc committed:
Compatibility with Python3.6

brejoc committed:
Added Windows fallback

Linux uses signal.alarm to just terminate the polling when a time
limit is hit.For Windows are are just counting the loop cycles.

salt-jenkins commented:
@brejoc, thanks for your PR! By analyzing the history of the files in this pull request, we identified @flavio, @meaksh and @mcalmer to be potential reviewers.

brejoc committed:
Fixed linting

* Python3 error for exception raising and
* Disabled linting for while-loop since the loop is broken via timeout.

brejoc committed:
Merge branch '2017.7' into improve-async-operation-handling-in-kubernetes-module

brejoc commented:
I need feedback with the linting:

1. from salt.ext.six.moves import range
```
17:36:40 ************* Module salt.modules.kubernetes
17:36:40 salt/modules/kubernetes.py:52: [W1699(incompatible-py3-code), ] Incompatible Python 3 code found. Proposed fix:
17:36:40  from salt.ext.six import iteritems
17:36:40  import salt.utils
17:36:40  import salt.utils.templates
17:36:40 +from salt.ext.six.moves import range
17:36:40  
17:36:40  try:
17:36:40      import kubernetes  # pylint: disable=import-self
```
If I add that, my linting complains, that this `salt.ext.six.moves.range` is not there. What to do?



2. Useless  else on loop
```
17:36:40 salt/modules/kubernetes.py:724: [W0120(useless-else-on-loop), delete_deployment] Else clause on loop without a break statement
17:36:44 + EC1=4
17:36:44 + pylint --rcfile=.testing.pylintrc --disable=W0232,E1002,W1307 tests/
17:36:44 + tee pylint-report-tests.xml
17:49:58 + EC2=0
17:49:58 + '[' 4 -ne 0 ']'
```
This should work and I would like to disable the linting for that. Is that possible?

brejoc commented:
@cachedout @rallytime Sorry for pushing this. Would anyone please be so kind to give feedback on this? Thanks!

meaksh commented on salt/modules/kubernetes.py:
```diff
@@ -692,7 +715,26 @@ def delete_deployment(name, namespace='default', **kwargs):
+                    break
+                else:
+                    sleep(1)
+        return mutable_api_response
```
Does it make sense to have the `log.warning` somewhere over here? Only when `mutable_api_response['code'] != 200` ?

meaksh commented on salt/modules/kubernetes.py:
```diff
@@ -78,6 +82,25 @@ def __virtual__():
+        finally:
+            signal.alarm(0)
+
+    _polling_time_limit = 30
```
What do you think about converting this `_polling_time_limit` into a constant like `POLLING_TIME_LIMIT` ?

brejoc replied:
Sure, why not!

meaksh reviewed (commented):
Hey @brejoc , thanks for this fix! I just add some feedback to the PR.

About the pylint issue, I'm not sure whether `salt.ext.six.moves.range` is compatible with `salt.ext.six.iteritems` and can be safely swapper or not.

Let's @cachedout or @rallytime put some light here, as currently I didn't get that `incompatible-py3-code` error with my linter.

brejoc commented on salt/modules/kubernetes.py:
```diff
@@ -692,7 +715,26 @@ def delete_deployment(name, namespace='default', **kwargs):
+                    else:
+                        mutable_api_response['code'] = 200
+            except TimeoutException:
+                pass
```
Yes, that's a good idea. But I think we can boil this down to one warning for both cases.

brejoc committed:
Code styling and added log message for timeout

brejoc commented:
Many thanks for the feedback @meaksh!

rallytime commented:
Hi @brejoc - My apologies for the slow reply about your lint questions.

1. For the range import, you should just be able to add the comment to disable the import error. There are a couple more you might need to add for your linter, as I am not sure which error yours is complaining about. [Here](https://github.com/saltstack/salt/blob/1ddf818b4102aba43cb4775fcb3db9f0b6bc2fb1/salt/modules/boto_vpc.py#L155) is an example of just the import error and [here](https://github.com/saltstack/salt/blob/1ddf818b4102aba43cb4775fcb3db9f0b6bc2fb1/salt/modules/network.py#L28) is a more involved error handling example.

2. For the useless else loop, you need to move the `# pylint: disable=useless-else-on-loop` comment off of the line it's currently on and onto the line that has the `else`. That should work nicely. :)

brejoc committed:
Added pylint-disable statements and import for salt.ext.six.moves.range

brejoc committed:
Merge branch '2017.7' into improve-async-operation-handling-in-kubernetes-module

brejoc committed:
Fixed test for delete_deployment

Due to implementation change, we need to mock the return value.

rallytime commented on salt/modules/kubernetes.py:
```diff
@@ -78,6 +83,25 @@ def __virtual__():
     return False, 'python kubernetes library not found'
 
 
+class TimeoutException(Exception):
```
Can you use one of the exceptions already defined in `salt.exceptions.py`?

brejoc replied:
I've looked through the list of exceptions and nothing seems to really fit.

rallytime replied:
Hrm. I am thinking that defining an exception class in the module itself might not be the best way to go moving forward.

What do you think @cachedout? Should something like this go in `salt/exceptions.py`? Is there a better name for this more general exception than the ones that are already in `salt/exceptions.py`?

cachedout replied:
I would vastly prefer that we keep all of our defined exceptions in one place. There should be no issue at all just moving this new one over there.

rallytime commented on salt/modules/kubernetes.py:
```diff
@@ -78,6 +83,25 @@ def __virtual__():
+    pass
+
+
+if not platform.startswith("win"):
```
Can you change this to use the salt utility (`salt.utils.is_windows()`)?

rallytime commented on salt/modules/kubernetes.py:
```diff
@@ -692,7 +716,30 @@ def delete_deployment(name, namespace='default', **kwargs):
             body=body)
-        return api_response.to_dict()
+        mutable_api_response = api_response.to_dict()
+        if not platform.startswith("win"):
```
Same comment here to use `salt.utils.is_windows()`.

rallytime commented on salt/modules/kubernetes.py:
```diff
@@ -692,7 +716,30 @@ def delete_deployment(name, namespace='default', **kwargs):
+                else:
+                    sleep(1)
+        if mutable_api_response['code'] != 200:
+            log.warning("Reached polling time limit. Deployment is not yet "
```
Can you change this string to use single quotes? [We like single quotes](https://docs.saltstack.com/en/latest/topics/development/conventions/style.html#single-quotes) instead of double quotes for strings.

rallytime reviewed (changes requested):
@brejoc I just had a couple more comments related to salt-specific things. I think this looks good once those are cleaned up. :)

brejoc committed:
Using salt method to identify MS Windows, single instead of double quotes

brejoc commented:
@rallytime Thanks for the feedback. Pushed a commit with the changes you suggested.

rallytime commented:
@brejoc There is a small lint error: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/14415/violations/file/salt/modules/kubernetes.py/

Once that is fixed up, I think the only thing left is the question about the exception class that I had above. Once @cachedout has had a chance to weigh in, we can get this in.

brejoc committed:
Removed unused sys import

brejoc commented:
Thanks @rallytime, I've fixed that and I'm waiting for the response from @cachedout.

brejoc committed:
Moved exception Salt core

The timeout exception is now part of exceptions.py and no longer solely
defined in the module.

brejoc committed:
Merge branch '2017.7' into improve-async-operation-handling-in-kubernetes-module

brejoc commented:
@cachedout Move the timeout exception to `exceptions.py` and updated the branch. Let me know if anything else is required. Thanks!

cachedout reviewed (approved):
I really like this approach. Nicely done!

Commit:
Merge branch '2017.7' into improve-async-operation-handling-in-kubernetes-module

brejoc commented on salt/modules/kubernetes.py:
```diff
@@ -78,6 +83,21 @@ def __virtual__():
     return False, 'python kubernetes library not found'
 
 
+if salt.utils.is_windows():
```
Oh, wait a second. This is wrong! This should be the other way around. Windows has *no* singnals.

brejoc commented on salt/modules/kubernetes.py:
```diff
@@ -692,7 +712,30 @@ def delete_deployment(name, namespace='default', **kwargs):
             body=body)
-        return api_response.to_dict()
+        mutable_api_response = api_response.to_dict()
+        if salt.utils.is_windows():
```
This should also be the other way around.

brejoc commented:
@cachedout @rallytime Please don't merge yet. The logic for the windows fallback is wrong. I have to fix that first. Sorry!

brejoc committed:
Fixed logic for windows fallback

Silly error - should have been the other way around.

brejoc commented:
@cachedout @rallytime Okay, once the tests are green, we should be good to go here.

cachedout commented:
@rallytime Re-review requested.

rallytime commented:
Thanks @brejoc. This looks great!

Test failures are not related.
